napari/layers/base/layer_event_handler.py

```diff
@@ -0,0 +1,37 @@
+import logging
+
+
+from ...utils.event_handler import EventHandler
+from napari.utils.base_interface import BaseInterface
+
+
+logger = logging.getLogger(__name__)
+
+
+class LayerEventHandler(EventHandler):
+    """
+    Event handler for layer specific events. Receives change events made from the data/controls/
+    or visual interface and updates all associated components.
+
+    Ex. ImageLayer components = [qt_image.py, vispy_image.py, image.py]
+    """
+
+    def __init__(self, component: BaseInterface = None):
+        super().__init__(component)
+
+    def on_change(self, event=None):
+        """
+        Process changes made from any interface
+        """
+        name = event.type
+        logger.debug(f"event: {name}")
+        try:  # until refactor on all layers is complete, not all events will have a value property
+            value = event.value
+        except AttributeError:
+            logger.debug(f"did not handle event {name}")
+            return
+
+        for component in self.components_to_update:
+            update_method_name = f"_on_{name}_change"
+            update_method = getattr(component, update_method_name)
+            update_method(value)
```

Comment:
One thing I'm wondering is - should the event handler be able to modify the value coming in from an event before calling the update methods?

This would maybe move the event handler towards being more of a controller where it actually did logic, but I'm having a hard time fitting in a couple of our more advanced use cases and it could simplify somethings elsewhere.

Take this example, the user types in a new name into the dialog box on the GUI, it triggers and event with that name, then the event arrives at the layer which sets the name, and at the layerlist/ viewer which has to check if the name is unique and if not coerces it to be unique, emitting a new event with the new name which then goes back to everyone. Here we now have two events and things could easily get out of sync. If the event handler itself did the checking then that might be simpler.

Basically I'm wondering if any place where we're putting logic in the `_on_*_change` method on the base layer that changes the value in a way that all the "listeners" will want to know, should probably go in the event handler. For example in the `_on_colormap_change` method we take the string and generate the actual colormap. This is needed in the vispy listener too (see @ttung comment here https://github.com/napari/napari/pull/1040/files#r391268444). If we did the transformation in the event handler then all our listeners could get the actual colormap - or maybe these transformations just shouldn't be living in the places they are living either!

Thoughts from @shanaxel42 @ttung @kne42 would be appreciated on this one

Reply:
I've given this a try here https://github.com/shanaxel42/napari/pull/2/commits/a9f182e886b360180f8ee619c117361e8fa3605f see comment here https://github.com/shanaxel42/napari/pull/2#issuecomment-647029262. I think it might work quite nicely, but curious what @napari/core-devs think